spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/actions/RewritePositionDeleteFilesSparkAction.java

```diff
@@ -133,55 +133,69 @@ public RewritePositionDeleteFiles.Result execute() {
+
+    CloseableIterable<PositionDeletesScanTask> tasks =
         CloseableIterable.transform(
             deletesTable.newBatchScan().ignoreResiduals().planFiles(),
             t -> (PositionDeletesScanTask) t);
 
     try {
       StructType partitionType = Partitioning.partitionType(table);
-      StructLikeMap<List<PositionDeletesScanTask>> filesByPartition =
-          StructLikeMap.create(partitionType);
-
-      for (PositionDeletesScanTask task : scanTasks) {
-        StructLike coerced = coercePartition(task, partitionType);
-
-        List<PositionDeletesScanTask> partitionTasks = filesByPartition.get(coerced);
-        if (partitionTasks == null) {
-          partitionTasks = Lists.newArrayList();
-        }
-        partitionTasks.add(task);
-        filesByPartition.put(coerced, partitionTasks);
-      }
-
-      StructLikeMap<List<List<PositionDeletesScanTask>>> fileGroupsByPartition =
-          StructLikeMap.create(partitionType);
 
-      filesByPartition.forEach(
-          (partition, partitionTasks) -> {
-            Iterable<List<PositionDeletesScanTask>> plannedFileGroups =
-                rewriter.planFileGroups(partitionTasks);
-            List<List<PositionDeletesScanTask>> groups = ImmutableList.copyOf(plannedFileGroups);
-            if (groups.size() > 0) {
-              fileGroupsByPartition.put(partition, groups);
-            }
-          });
+      StructLikeMap<List<PositionDeletesScanTask>> filesPerPartition =
+          filesByPartition(partitionType, tasks);
 
-      return fileGroupsByPartition;
+      return fileGroupsByPartition(partitionType, filesPerPartition);
     } finally {
       try {
-        scanTasks.close();
+        tasks.close();
       } catch (IOException io) {
         LOG.error("Cannot properly close file iterable while planning for rewrite", io);
       }
     }
   }
 
-  @VisibleForTesting
-  RewritePositionDeletesGroup rewriteDeleteFiles(
+  private StructLikeMap<List<PositionDeletesScanTask>> filesByPartition(
+      StructType partitionType, Iterable<PositionDeletesScanTask> tasks) {
+    StructLikeMap<List<PositionDeletesScanTask>> filesByPartition =
+        StructLikeMap.create(partitionType);
+
+    for (PositionDeletesScanTask task : tasks) {
+      StructLike coerced = coercePartition(task, partitionType);
+
+      List<PositionDeletesScanTask> partitionTasks = filesByPartition.get(coerced);
+      if (partitionTasks == null) {
+        partitionTasks = Lists.newArrayList();
+      }
+      partitionTasks.add(task);
+      filesByPartition.put(coerced, partitionTasks);
+    }
+
+    return filesByPartition;
+  }
+
+  private StructLikeMap<List<List<PositionDeletesScanTask>>> fileGroupsByPartition(
+      StructType partitionType, StructLikeMap<List<PositionDeletesScanTask>> filesByPartition) {
+    StructLikeMap<List<List<PositionDeletesScanTask>>> fileGroupsByPartition =
+        StructLikeMap.create(partitionType);
+
+    filesByPartition.forEach(
```

Comment:
question: It seems we would benefit from an easy way to transform `StructLikeMap` values. It is used here and below when we construct the context. Would adding something like this to `StructLikeMap` be an overkill?

```
public <U> StructLikeMap<U> transformValues(Function<T, U> func) {
  StructLikeMap<U> map = create(type);

  for (Entry<StructLikeWrapper, T> entry : wrapperMap.entrySet()) {
    U newValue = func.apply(entry.getValue());
    map.put(entry.getKey().get(), newValue);
  }

  return map;
}
```

Then the entire method `fileGroupsByPartition` can be expressed in one line:

```
tasksByPartition.transformValues(this::planFileGroups);

...

private List<List<PositionDeletesScanTask>> planFileGroups(List<PositionDeletesScanTask> tasks) {
  return ImmutableList.copyOf(rewriter.planFileGroups(tasks));
}
```

And `RewriteExuctionContext` would call:

```
this.numGroupsByPartition = groupsByPartition.transformValues(List::size);
```

We would need to filter out empty file group lists but we can do that in `toGroupStream`. I am not sure it is worth it so I leave it up to you.

Reply:
I like this idea, implemented it.  I was wondering initially as its not the pattern of the static method like CloseableIterable.transform(), but I think putting on instance level makes the code shorter and even more functional and composable.